### ISSUE TITLE
Dot (.) character of URI issue

### ISSUE THREAD

**Issue by allenk** [civetweb/civetweb#816]
I use "mg_set_request_handler" to hook the specified URI to my handler.

for example,
"curl https://127.0.0.1/%2e/SECURE_API/neverseen.txt"
and
"curl https://127.0.0.1/./SECURE_API/./neverseen.txt"

If I use "." in my url, it looks like curl will remove "." from request url.
If I use "%2e", it will send to civetweb directly and the conn->request_uri will be
"/./SECURE_API/./neverseen.txt".

For the case, my request handler is never be matched.
The worse is that my URI filter will be bypass using "%2e".

I see function "remove_double_dots_and_double_slashes" will try to remove "..".
Maybe we should remove ".", too.

**Comment by bel2125** [comment]
I did some quick research, and it seems resources just named "." (dot) are problematic:
https://stackoverflow.com/questions/3856693/a-url-resource-that-is-a-dot-2e

From the "combined answer" there:
> §2.3 (of [RFC 3986](https://www.ietf.org/rfc/rfc3986.txt)) says that "." (period) is an unreserved character and that "URIs that differ in the replacement of an unreserved character with its corresponding percent-encoded US-ASCII octet are equivalent". 

Thus, 
"curl https://127.0.0.1/%2e/SECURE_API/neverseen.txt" should be "/./SECURE_API/neverseen.txt"
and
"curl https://127.0.0.1/./SECURE_API/./neverseen.txt" should be "/./SECURE_API/./neverseen.txt"

CivetWeb does this in the first step of request processing, before any handler is called:
https://github.com/civetweb/civetweb/blob/7d82fadf366a76d41a16660c016378de48b54252/src/civetweb.c#L13797-L13801
(there is a configuration [`decode_url`](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#decode_url-yes) to disable decoding ... I think this would violate the RFC section mentioned above but might be useful in a proxy configuration).

> I see function "remove_double_dots_and_double_slashes" will try to remove "..".
> Maybe we should remove ".", too.

This function is just the next and final step of URI "pre-processing":
https://github.com/civetweb/civetweb/blob/7d82fadf366a76d41a16660c016378de48b54252/src/civetweb.c#L13803-L13805
It's purely meant as a protection before accessing the server file system. From that point of view, using "/./myFile.txt" in not problematic, since it's just the same as "/myFile.txt" anyway, while "/../fileInParentDirectory.txt" is something different.
An edge case (if you use a "NO_FILESYSTEM" define): If you use a request handlers for *everything* and never use an automatic "URI -->file path" translation, using '/../something" would not be a problem.

The 'remove_double_dots_and_double_slashes' is "old code" that has not been touched for years ... when looking at it now, I have two tendencies:
1) either replace it completely by something like 
`if (is_double_dot_in_uri) { send(403 Forbidden); return;}`
2) or replacing it by a `remove_dot_segments` according to https://tools.ietf.org/html/rfc3986#section-5.2.4

Ad 2) I guess this `remove_dot_segments` is exactly what "curl" did when it removes "." from your request url.
Ad 1) One could consider "strange combinations of slashes and dots" as an attempted attack / search for an exploit and "just block everything like that". Maybe some components (proxies, caches, other servers, ...) might have already done that, and this could cause problems to everybody trying to use resources with names that do not start/end with alphanumeric characters.

**Comment by allenk** [comment]
Yes, Ad 2) seems to work well after my test.
I try to add some test code in remove_double_dots_and_double_slashes().
follow ".." and try to test "."

```
    while ((s[0] == '.')) {
        s++;
    }
    ......
    else if ((s[0] == '.')) {
        s++;
    }    
```

Now, I don't need use "**/SECURE_API/neverseen.txt" or "/SECURE_API/**/neverseen.txt" to filter all possible URI patterns. "/SECURE_API/neverseen.txt" will be triggered correctly.

**Comment by allenk** [comment]
Unfortunately.
`
https://127.0.0.1/SECURE_API%2e/neverseen.txt
->
https://127.0.0.1/SECURE_AP./neverseen.txt
`
This URI can access neverseen.txt.
I expect it will send Error 404: Not Found.

It seems that more consideration is needed.

**Comment by bel2125** [comment]
> `https://127.0.0.1/SECURE_API%2e/neverseen.txt -> https://127.0.0.1/SECURE_AP./neverseen.txt`
This URI can access neverseen.txt.

What did you register as handler pattern?
What did you pass to `curl`?

**Comment by allenk** [comment]
curl commands:
`
curl https://127.0.0.1/SECURE_API%2e/neverseen.txt
`

I try
```
mg_set_request_handler(ctx, "/SECURE_API/neverseen.txt", RESTIMPL_Debug_Secure_URI, 0);
mg_set_request_handler(ctx, "**/SECURE_API/neverseen.txt", RESTIMPL_Debug_Secure_URI, 0);
mg_set_request_handler(ctx, "/SECURE_API/**neverseen.txt", RESTIMPL_Debug_Secure_URI, 0);
```

It cannot match to "/SECURE_API./neverseen.txt".
It look like "/SECURE_API./neverseen.txt" could access my "webroot//SECURE_API/neverseen.txt".

> "neverseen.txt" is a dummy file to verify the secure access.

Try to remove all "." from localurl but lots documents of my webroot are not accessible.

This pushes me to use 
`callbacks.begin_request`
filter all my URLs.

**Comment by allenk** [comment]
I implement:
```
callbacks.begin_request = [](struct mg_connection *conn) {
    ......
    remove_all_dots(pSecureURI);
    remove_double_slashes(pSecureURI);
    ......
    if (!mg_strncasecmp(...)) {
        // do my stuff for the URI.
        // return 404 or 401 ...
    }
    return 0;
};
```
With this design, URIs are only affected by my filters, and file resources are still accessible.
Of course, the URI matching mechanism remains a problem.

**Comment by bel2125** [comment]
Would you try this implementation?
```C
static void
remove_double_dots_and_double_slashes(char *s)
{
	/* Algorithm "remove_dot_segments" from https://tools.ietf.org/html/rfc3986#section-5.2.4 */

	/* Step 1:
	 * The input buffer "in" is initialized.
	 * The output buffer "s" is initialized to the empty string.
	 */
	char *s_copy = mg_strdup(s);
	char *in = s_copy;
	char *out_end = s;
	*s = 0;

	/* Step 2:
	 * While the input buffer is not empty, loop as follows:
	 */
	while (*in) {
		/* Step 2a:
		 * If the input buffer begins with a prefix of "../" or "./",
         * then remove that prefix from the input buffer;
		 */
		if (!strncmp(in, "../", 3)) {
			in += 3;
		} 
		else if (!strncmp(in, "./", 2)) {
			in += 2;
		} 
		/* otherwise */
		/* Step 2b:
		 * if the input buffer begins with a prefix of "/./" or "/.",
         * where "." is a complete path segment, then replace that
         * prefix with "/" in the input buffer;
		 */
		else if (!strncmp(in, "/./", 3)) {
			in += 2;
		}
		else if (!strcmp(in, "/.")) {
			in[1] = 0;
		}
		/* otherwise */
		/* Step 2c:
		 * if the input buffer begins with a prefix of "/../" or "/..",
		 * where ".." is a complete path segment, then replace that
		 * prefix with "/" in the input buffer and remove the last
		 * segment and its preceding "/" (if any) from the output
		 * buffer;
		 */
		else if (!strncmp(in, "/../", 4)) {
			in += 3;
			if (s != out_end) {
				/* remove last segment */
				do {
					out_end--;
					*out_end = 0;
				} while ((s != out_end) && (*out_end != '/'));
			}
		}
		else if (!strcmp(in, "/..")) {
			in[1] = 0;
			if (s != out_end) {
				/* remove last segment */
				do {
					out_end--;
					*out_end = 0;
				} while ((s != out_end) && (*out_end != '/'));
			}
		}
		/* otherwise */
		/* Step 2d:
		 * if the input buffer consists only of "." or "..", then remove
		 * that from the input buffer;
		 */
		else if (!strcmp(in, ".") || !strcmp(in, ".")) {
			*in = 0;
		}
		/* otherwise */
		/* Step 2d:
		 * move the first path segment in the input buffer to the end of
		 * the output buffer, including the initial "/" character (if
		 * any) and any subsequent characters up to, but not including,
		 * the next "/" character or the end of the input buffer.
		 */
		else {
			do {
				*out_end = *in;
				out_end++;
				in++;
			} while ((*in != 0) && (*in != '/'));
		}
	}

	/* Step 3:
	 * Finally, the output buffer "s" is returned as the result of
     * remove_dot_segments.
	 */
	/* Terminate output */
	*out_end = 0;

	/* Free temporary copies */
	mg_free(s_copy);
}

```

This is according to the RFC (unless I made a mistake).

**Comment by allenk** [comment]
I have done the test. It works for all my test cases and request handler is triggered correctly.
But, one of my test case is not passed.

`https://127.0.0.1/%2e/SECURE_API%2e/neverseen.txt`

**"/SECURE_API./neverseen.txt"** is accepted by filesystem access
so I think this should be converted to
**"/SECURE_API/neverseen.txt"**

I update your implementation and then all test cases are passed.
```c++
        ......
        /* otherwise */
        /* Step 2d:
         * move the first path segment in the input buffer to the end of
         * the output buffer, including the initial "/" character (if
         * any) and any subsequent characters up to, but not including,
         * the next "/" character or the end of the input buffer.
         */
        else {
            do {
                *out_end = *in;
                out_end++;
                in++;
                // skip '.' of './'
                if (!strncmp(in, "./", 2)) {
                    in++;
                }
            } while ((*in != 0) && (*in != '/'));
        }
        ......
```
skip "." in "./" will fix the problem.

**Comment by bel2125** [comment]
Before I updated my post, I had the following bug in 2d:

`else if (!strcmp(in, ".") || !strcmp(in, "..")) {` <-- now
`else if (!strcmp(in, ".") || !strcmp(in, ".")) {` <-- before

... ah, I used 2d twice

**Comment by bel2125** [comment]
So, the remaining problem is, there is a difference between 
"name" and "name." (dot at the end of a file name with no file extension).
I wonder if this is a "Windows only" problem ... got to check if Linux allows a file with a dot at the end of the file name and if it's different to the same name without dot

**Comment by bel2125** [comment]
.. seems to be another Windows vs. Linux topic:
Linux allows "x" and "x.", and they are two different files.
For Windows "x" and "x." is the same file (or folder in your case).

My approach would be not to modify the RFC algorithm, but to do another (Windows specific) post-processing at the end. There is already a (Windows specific) backslash to slash preprocessing for this algorithm.

**Comment by allenk** [comment]
I think it might be a "Windows only".
I do following test:
`c:\ type c:\test\neverseen.txt`
or
`c:\ type c:\test.\neverseen.txt`
(dir command get the same result.)

Then I try it again with Ubuntu 18.04:
`cat /home/allen/test./neverseen.txt`
`cat: /home/allen/test./neverseen.txt: No such file or directory`

Interestingly, if I use cat in msys, I get the same result of Ubuntu.

**Comment by bel2125** [comment]
... new end of the function:
```C
	/* Step 3:
	 * Finally, the output buffer is returned as the result of
	 * remove_dot_segments.
	 */
	/* Terminate output */
	*out_end = 0;

	/* For Windows, the files/folders "x" and "x." (with a dot but without
	 * extension) are identical. Replace all "./" by "/" and remove a "." at the
	 * end.
	 */
	out_end = out_begin;
	while (*out_end) {
		if ((*out_end == '.') && ((out_end[1] == '/') || (out_end[1] == 0))) {
			char *r = out_end;
			do {
				r[0] = r[1];
				r++;
			} while (r[0]!=0);
		}
		out_end++;
	}

	/* Free temporary copies */
	mg_free(in_copy);
}
```

**Comment by allenk** [comment]
I can confirm it works for all cases. Thanks.

**Comment by bel2125** [comment]
I just have to update the test cases, before this change will be merged.

**Comment by xtne6f** [comment]
You have to revert fa3f169ff396e142743a2858e008673e1d1afbce and 754441572413050f6ea64519d46f32ec8546c5b4 for now, they can cause directory traversal by "/.../".

**Comment by allenk** [comment]
I have solved the directory traversal issue and fix another windows issue that "//" will treat as "/" on Windows filesystem. Please search my tag 'allen k' to find my modification.

```c++
static void
remove_dot_segments(char *inout)
{
	/* Windows backend protection
	 * (https://tools.ietf.org/html/rfc3986#section-7.3): Replace backslash in
	 * URI by slash */
	char *in_copy = mg_strdup(inout);
	char *out_begin = inout;
	char *out_end = inout;
	char *in = in_copy;
	char *in_end = in_copy;

        // allen k, convert \\ to / and skip all duplicated /
	while (*in) {
		if (*in == '\\') {
			*in = '/';
		}
		if (in_end == in_copy || in[0] != '/' || in_end[-1] != '/') {
			*in_end++ = *in;
		}
		in++;
	}
	in_end[0] = 0;

	/* Algorithm "remove_dot_segments" from
	 * https://tools.ietf.org/html/rfc3986#section-5.2.4 */
	/* Step 1:
	 * The input buffer is initialized.
	 * The output buffer is initialized to the empty string.
	 */
	in = in_copy;

	/* Step 2:
	 * While the input buffer is not empty, loop as follows:
	 */
	while (*in) {
		/* Step 2a:
		 * If the input buffer begins with a prefix of "../" or "./",
		 * then remove that prefix from the input buffer;
		 */
		if (!strncmp(in, "../", 3)) {
			in += 3;
		} else if (!strncmp(in, "./", 2)) {
			in += 2;
		}

		/* otherwise */
		/* Step 2b:
		 * if the input buffer begins with a prefix of "/./" or "/.",
		 * where "." is a complete path segment, then replace that
		 * prefix with "/" in the input buffer;
		 */
		else if (!strncmp(in, "/./", 3)) {
			in += 2;
		} else if (!strcmp(in, "/.")) {
			in[1] = 0;
		}
		/* otherwise */
		/* Step 2c:
		 * if the input buffer begins with a prefix of "/../" or "/..",
		 * where ".." is a complete path segment, then replace that
		 * prefix with "/" in the input buffer and remove the last
		 * segment and its preceding "/" (if any) from the output
		 * buffer;
		 */
		else if (!strncmp(in, "/../", 4)) {
			in += 3;
			if (out_begin != out_end) {
				/* remove last segment */
				do {
					out_end--;
					*out_end = 0;
				} while ((out_begin != out_end) && (*out_end != '/'));
			}
		} else if (!strcmp(in, "/..")) {
			in[1] = 0;
			if (out_begin != out_end) {
				/* remove last segment */
				do {
					out_end--;
					*out_end = 0;
				} while ((out_begin != out_end) && (*out_end != '/'));
			}
                // allen k, convert /.. to / and .. to .
		} else if (!strncmp(in, "/..", 3)) {
			*out_end = *in;
			out_end++;
			in++;
		} else if (!strncmp(in, "..", 2)) {
			in++;
		}

		/* otherwise */
		/* Step 2d:
		 * if the input buffer consists only of "." or "..", then remove
		 * that from the input buffer;
		 */
		else if (!strcmp(in, ".") || !strcmp(in, "..")) {
			*in = 0;
		}
		/* otherwise */
		/* Step 2e:
		 * move the first path segment in the input buffer to the end of
		 * the output buffer, including the initial "/" character (if
		 * any) and any subsequent characters up to, but not including,
		 * the next "/" character or the end of the input buffer.
		 */
		else {
			do {
				*out_end = *in;
				out_end++;
				in++;
			} while ((*in != 0) && (*in != '/'));
		}
	}

	/* Step 3:
	 * Finally, the output buffer is returned as the result of
	 * remove_dot_segments.
	 */
	/* Terminate output */
	*out_end = 0;

	/* For Windows, the files/folders "x" and "x." (with a dot but without
	 * extension) are identical. Replace all "./" by "/" and remove a "." at the
	 * end.
	 */
	out_end = out_begin;
	while (*out_end) {
		if ((*out_end == '.') && ((out_end[1] == '/') || (out_end[1] == 0))) {
			char *r = out_end;
			do {
				r[0] = r[1];
				r++;
			} while (r[0] != 0);
		}
		out_end++;
	}

	/* Free temporary copies */
	mg_free(in_copy);
}
```

**Comment by xtne6f** [comment]
I can't find the tag in your repository.. Where is?

**Comment by allenk** [comment]
> 
> 
> I can't find the tag in your repository.. Where is?

No, I just mean you can find it from my comments. I don't a repo. for that.

**Comment by bel2125** [comment]
> You have to revert fa3f169 and 7544415 for now, they can cause directory traversal by "/.../".

Yes, this has to be solved quickly ... a hot fix would be drop all paths containing 3 dots.

**Comment by bel2125** [comment]
This quick fix 16b72c1a325d6c8c3f1549a9ba780d1f845103d1 replaces all "........................../" of arbitrary length by repeating the replace "./" --> "/" operation as long as there is any "./" 

```C
	/* For Windows, the files/folders "x" and "x." (with a dot but without
	 * extension) are identical. Replace all "./" by "/" and remove a "." at the
	 * end.
	 */
	do {
		replaced = 0;
		out_end = out_begin;
		while (*out_end) {
			if ((*out_end == '.')
			    && ((out_end[1] == '/') || (out_end[1] == 0))) {
				char *r = out_end;
				do {
					r[0] = r[1];
					r++;
					replaced = 1;
				} while (r[0] != 0);
			}
			out_end++;
		}
	} while (replaced);
```

**Comment by allenk** [comment]
Just done a quick test on 6fad4ba8d784c925ee279ce1409535e405cccb8c and haven't found any problems. I will send it to vulnerability testing software and will get the result back soon. Thanks!

**Comment by bel2125** [comment]
Any news from vulnerability testing? I haven't found any problem, but maybe someone else ...
We must be sure, there is no security bug in the release?

**Comment by bel2125** [comment]
Since we want to release the new version since quite a while (#795): is there any vulnerability we do not know yet? Or can we close this issue?

**Comment by allenk** [comment]
> Since we want to release the new version since quite a while (#795): is there any vulnerability we do not know yet? Or can we close this issue?

We recently completed testing.  In the report of the testing from security test company, the vulnerability has not been reproduced anymore, so that we can close the issue ! Thanks!

**Comment by xtne6f** [comment]
Please let me confirm one thing to @allenk. the following code:
```
do {
    out_end--;
    *out_end = 0;
} while ((out_begin != out_end) && (*out_end != '/'));
```
has a bug or not ? I think the `*out_end = 0;` is unneeded. The current code converts "/foo/bar/baz/../qux.txt" into "/qux.txt". This will not cause security issues, but...

**Comment by bel2125** [comment]
These are the current test cases:

https://github.com/civetweb/civetweb/blob/2261bfc7a15b66705bf82905d6f8a32324bf9c6e/unittest/private.c#L331-L376

But there seems to be nothing like "/foo/bar/baz/../qux.txt".

Removing the `*out_end = 0;` ... Probably it should be moved out if the loop, so the string is always terminated.

```C
			if (out_begin != out_end) {
				/* remove last segment */
				do {
					out_end--;
					//*out_end = 0;   <-- CUT
				} while ((out_begin != out_end) && (*out_end != '/'));
				*out_end = 0;  // <-- PASTE
			}
```

**Comment by xtne6f** [comment]
`*out_end = 0;  // <-- PASTE` would be not necessary since "Step 3" terminates output finally.

**Comment by bel2125** [comment]
I just completed all tests with the `*out_end = 0; // <-- PASTE` ... I did no want to wait for "Step 3" since "PASTE" is inside the `while` loop and "Step 3" only after this loop ... but after checking all conditions in the while loop, I think you are right.

**Comment by xtne6f** [comment]
Thank you, I will re-close this issue.

The PASTE line should have no side effects if you have no plans to share the input/output buffer. Actually, `remove_dot_segments()` does not need `strdup()` because it never do "append characters" operations.
